### ISSUE TITLE
NPE during processing nested arrays

### ISSUE THREAD

**Issue by abracadv8** [mapstruct/mapstruct#734]
My apologies in advance, as this is my first time collaborating with Github.

I'll create a fork to link to the relevant test case, but I would think this scenario should either be implicitly handled or at least given a proper error message.

Fork: https://github.com/abracadv8/mapstruct/tree/feature/copyConstructorNpe/processor/src/test/java/org/mapstruct/ap/test/bugs/_734

One class contains a simple array of target types (ArrayContainerTarget) and the other class (ArrayContainerSource) contains a wrapper class containing an ArrayList of source types (similar to how some tools generate arrays from XSDs or WSDLs).

```
public class ArrayContainerTarget {
    private Target[] targets;
}

public class ArrayContainerSource {
    private ArrayOfSource sources;
}

public class ArrayOfSource {
    private List<Source> sources;
}
```

The mapping then looks as follows: 

```
    public abstract Source map( Target target );
    public abstract Target map( Source target );
    public abstract Source[] map1( List<Target> target );
    @Mappings( {
        @Mapping( source = "sources.sources", target = "targets" )
    } )
    public abstract ArrayContainerTarget map( ArrayContainerSource source );
```

The processor ends up failing with an NPE at this line because `targetType.getTypeElement()` is null.

```
Internal error in the mapping processor: java.lang.NullPointerException  
    at org.mapstruct.ap.internal.processor.creation.MappingResolverImpl$ResolvingAttempt.hasCompatibleCopyConstructor(MappingResolverImpl.java:549)
```

I would expect either:

1. A message telling me that I am missing some mapping.
1. It be able to implicitly convert a List<> to an array [].

**Comment by abracadv8** [comment]
I created some sysouts in the MappingResolverImpl just to see what is going on.  

It may not be elegant, but simply returning if there is no TypeElement may work; it seems to produce a valid mapping because one of the other resolvers correctly creates a method for it.

Example: https://github.com/abracadv8/mapstruct/blob/feature/copyConstructorNpe/processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java#L552

```
if ( targetType.getTypeElement() == null ) {
    // FIXME
    System.out.println( "\n--- targetType.getTypeElement() == null, bailing out (false)\n" );
    // Assuming it is false seems to produce a correct mapping
    //return false;
}
```

**Comment by gunnarmorling** [comment]
Agreed this should be supported OOTB (and indeed I would have assumed it works). Need to dig a bit further to grok the details. If you think you've got a solution, we'd gladly apply your PR.

**Comment by abracadv8** [comment]
Sure, thanks.  Take a look and let me know your thoughts. 

If the fix looks good to you, I'll clean up the statements and debugging and submit a PR.

**Comment by gunnarmorling** [comment]
@abracadv8, I just pushed a fix for a very similar bug to master.

Could you fetch the latest master, build MapStruct yourself and see whether that fixes the error for you? Essentially we should only ever get to `hasCompatibleCopyConstructor()` if source and target type either are both a `Collection` or both a `Map` sub-type, which isn't the case for arrays. That should be fixed now.

As the direct assignment from list to array doesn't work, MapStruct will then generate a private method for doing this mapping, as you suggested.

Let me know how it goes :)

**Comment by abracadv8** [comment]
Yes!  That code appears to have solved the issue with the test case I created.

The Impl looks to have all the correct mappings for the nested lists and the private/protected methods to convert the nested classes.

Thanks!

**Comment by gunnarmorling** [comment]
Cool, I'll close this issue then. Thanks for the quick test!

**Comment by gunnarmorling** [comment]
Fixed by #729.
